src/console_gui.cpp

```diff
@@ -235,9 +187,12 @@ struct IConsoleWindow : Window
 	{
 		if (this->truncate_timer.CountElapsed(delta_ms) == 0) return;
 
-		if (IConsoleLine::Truncate() &&
-				(IConsoleWindow::scroll > IConsoleLine::size)) {
-			IConsoleWindow::scroll = std::max(0, IConsoleLine::size - (this->height / this->line_height) + 1);
+		assert(this->height >= 0 && this->line_height > 0);
+		size_t visible_lines = (size_t)(this->height / this->line_height);
+
+		if (TruncateBuffer() && IConsoleWindow::scroll + visible_lines > _iconsole_buffer.size()) {
+			size_t max_scroll = (visible_lines > _iconsole_buffer.size()) ? 0 : _iconsole_buffer.size() + 1 - visible_lines;
+			IConsoleWindow::scroll = std::min<size_t>(IConsoleWindow::scroll, max_scroll);
```

Comment:
I'm not entirely sure if this is correct, and I can't think of a good way to exercise the truncation code right now.

Reply:
lgtm